### PR TITLE
Feat: Restoring instruments list while switching back and forth between in instruments and templates

### PR DESCRIPTION
**Pull request by chilkaditya**
Resolves: #14342

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [ ] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

File: src/instrumentsscene/qml/MuseScore/InstrumentsScene/InstrumentsDialog.qml [hunk 1]
```diff
@@ -25,10 +25,12 @@ import QtQuick.Layouts 1.15
 import Muse.Ui 1.0
 import Muse.UiComponents 1.0
 import MuseScore.InstrumentsScene 1.0
+import MuseScore.Project 1.0
 
 StyledDialogView {
     id: root
 
+    required property InstrumentsOnScoreListModel instrumentsOnScoreModel
```
Comment (cbjeukendrup) [inline src/instrumentsscene/qml/MuseScore/InstrumentsScene/InstrumentsDialog.qml:33]: This will not work: `InstrumentsDialog` is not instantiated directly from QML, but as a dialog. Therefore, it won't be possible to pass anything to this required property, and the dialog will fail to open. 

Instead, you will need to instantiate an `InstrumentsOnScoreListModel` here inside `InstrumentsDialog`, just like you do in `ChooseInstrumentsAndTemplatesPage`.

File: src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml [hunk 2]
```diff
@@ -26,9 +26,11 @@ import QtQuick.Controls 2.12
...
 import Muse.UiComponents 1.0
 import MuseScore.InstrumentsScene 1.0
+import MuseScore.Project 1.0
 
 Item {
     id: root
+    required property InstrumentsOnScoreListModel instrumentsOnScoreModel
 
     readonly property bool hasInstruments: instrumentsOnScoreModel.count > 0
     readonly property alias isMovingUpAvailable: instrumentsOnScoreModel.isMovingUpAvailable
```
Comment (cbjeukendrup) [inline src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml:36]: The errors you are seeing are caused by these two aliases:
```
13:08:15.122 | DEBUG | main_thread     | createDialog    | [qml] failed create component: ../../MuseScore/Project/NewScoreDialog.qml, err: qrc:/qml/MuseScore/Project/NewScoreDialog.qml:87 Type ChooseInstrumentsAndTemplatesPage unavailable
qrc:/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml:174 Type ChooseInstrumentsPage unavailable
qrc:/qml/MuseScore/InstrumentsScene/ChooseInstrumentsPage.qml:196 Type InstrumentsOnScoreView unavailable
qrc:/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml:36 Invalid alias reference. Unable to find id "instrumentsOnScoreModel"
```
Indeed, `instrumentsOnScoreModel` is not an id anymore in this context, but a property. That means that it cannot be used for `alias` properties. 
Fortunately, the solution is easy: replace `alias` with `bool`.
But there is also another solution, that may be even better. If you look where `isMovingUpAvailable` and `isMovingDownAvailable` are used, you see that they are only used in `ChooseInstrumentsPage.qml`. But now that file has direct access to the `instrumentsOnScoreModel`, so there is no need anymore to take these properties via `InstrumentsOnScoreView`. 
So, the solution I would recommend is:
- remove these two alias properties from InstrumentsOnScoreView
- In ChooseInstrumentsPage, replace `instrumentsOnScoreView.isMovingUpAvailable` with `root.instrumentsOnScoreModel.isMovingUpAvailable`, and similar for `isMovingDownAvailable`.

While at it, you'll probably notice that InstrumentsOnScoreView contains a lot of functions (`instruments()`, `moveSelectedInstrumentsDown()`, etc) that are only for the purpose of exposing information of `instrumentsOnScoreModel`. But again, that is not necessary anymore, because now `instrumentsOnScoreModel` is available directly. So, it would be great to remove those functions, and use `instrumentsOnScoreModel` directly in ChooseInstrumentsPage.

File: src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml [hunk 3]
```diff
@@ -60,9 +62,7 @@ Item {
         instrumentsView.positionViewAtEnd()
     }
 
-    InstrumentsOnScoreListModel {
-        id: instrumentsOnScoreModel
-    }
+
```
Comment (cbjeukendrup) [inline src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml:66]: Make sure to delete unnecessary whitespace; one empty line is enough :)

Comment (chilkaditya) [reply src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml:66]: Ok, got it

File: src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml (hunk 2 above)
Comment (chilkaditya) [reply src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml:36]: It is kind of frustrating for me that after so much of  instructions I could not find out the correct solution  but I am listing my progress.  @cbjeukendrup  As per your instructions I update the InstrumentsDialog.qml, ChooseInstrumentsPage.qml and InstrumentsOnScoreView.qml. And code of those files are as follows:
InstrumentsDialog.qml
```qml
StyledDialogView {
    ......
    InstrumentsOnScoreListModel {
        id: instrumentsOnScoreModel
    }
    ......
}
```
ChooseInstrumentsPage.qml
```qml
function instruments() {
        if (root.canSelectMultipleInstruments) {
            return root.instrumentsOnScoreModel.instruments() // changed code
        }

        return [ instrumentsModel.selectedInstrument ]
}
function currentOrder() {
        if (root.canSelectMultipleInstruments) {
            return root.instrumentsOnScoreModel.currentOrder() // changed code
        }

        return undefined
}
function addSelectedInstrumentsToScore() {
        root.instrumentsOnScoreModel.addInstruments(instrumentsModel.selectedInstrumentIdList()) // changed code

        Qt.callLater(root.instrumentsOnScoreModel.scrollViewToEnd) // changed code
}
FlatButton {
        enabled: root.instrumentsOnScoreModel.isMovingUpAvailable // changed code
}
FlatButton {
        enabled: root.instrumentsOnScoreModel.isMovingDownAvailable // changed code
}
```
And I deleted two alice properties isMovingUpAvailable, isMovingDownAvailable and all those functions instruments(), currentOrder(), addInstruments, moveSelectedInstrumentsUp(), moveSelectedInstrumentsDown(), scrollViewToEnd() from the InstrumentsOnScoreView.qml.
After this Newscore creation dialog is opening but i can not add instruments to the list and It is showing in application output that 
error: qrc:/qml/MuseScore/InstrumentsScene/ChooseInstrumentsPage.qml:85: TypeError: Cannot read property 'addInstruments' of null

File: src/project/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml [hunk 4]
```diff
@@ -27,10 +27,11 @@ import Muse.Ui 1.0
 import Muse.UiComponents 1.0
 import MuseScore.Project 1.0
 import MuseScore.InstrumentsScene 1.0
-
 Item {
     id: root
 
+    required property InstrumentsOnScoreListModel instrumentsOnScoreModel
```
Comment (cbjeukendrup) [inline src/project/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml:33]: This property should not be here. In this file, the `InstrumentsOnScoreListModel` is instantiated, so there should be no property. Properties are usually only used to pass information from outside the file into the file, or to expose information from inside the file to outside the file.

Comment (chilkaditya) [reply src/project/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml:33]: Ok, got it.

File: src/project/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml [hunk 5]
```diff
@@ -27,10 +27,11 @@ import Muse.Ui 1.0
 import Muse.UiComponents 1.0
 import MuseScore.Project 1.0
 import MuseScore.InstrumentsScene 1.0
-
```
Comment (cbjeukendrup) [inline src/project/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml:30]: Unnecessary deletion

File: src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml (hunk 2 above)
Comment (chilkaditya) [reply src/instrumentsscene/qml/MuseScore/InstrumentsScene/internal/InstrumentsOnScoreView.qml:36]: Yes Yes, it is there in the InstrumentsOnScoreView.qml ,I mistakenly add in the comment.

File: src/project/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml [hunk 6]
```diff
@@ -168,6 +173,7 @@ Item {
 
         ChooseInstrumentsPage {
             navigationSection: root.navigationSection
+            instrumentsOnScoreModel: instrumentsOnScoreModel
```
[the remark this answers is not in the document]
Comment (chilkaditya) [reply src/project/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml]: I commit my progress. Not getting any error but not restoring the instruments list while switching between two. I think we have to do some changes in **addSelectedInstrumentsToScore()** function in **ChooseInstrumentsPage.qml**. There we call **addInstruments()** that is not exist anymore.

[the remark this answers is not in the document]
Comment (chilkaditya) [reply src/project/qml/MuseScore/Project/internal/NewScore/ChooseInstrumentsAndTemplatesPage.qml]: Please ignore comments as of now, I will delete those in my final and working commit.

**Comment by cbjeukendrup** [comment]
@chilkaditya Sorry for the delay, some other work came in between, but now I've found the reason that it was still not working, see the last third commit that I added. Solving it was in the end not difficult, but finding it was a bit tricky indeed! So I think this PR is now ready. Thanks for your work!

**Review by cbjeukendrup — APPROVED** [review]

**Comment by chilkaditya** [comment]
> @chilkaditya Sorry for the delay, some other work came in between, but now I've found the reason that it was still not working, see the last third commit that I added. Solving it was in the end not difficult, but finding it was a bit tricky indeed! So I think this PR is now ready. Thanks for your work!

Ok, I understand what was the problem. Thank you again for your help.

**Comment by zacjansheski** [comment]
Tested on MacOS 14, Windows 11, Ubuntu 22.04.3. Approved
https://github.com/musescore/MuseScore/issues/14342 FIXED
